### DISCUSSION TITLE
Avoid consuming input Vecs?

### DISCUSSION THREAD

**Discussion by jrbrodie77** [NicolasKlenert/enterpolation#19] [Q&A] [answered]
This is definitely a newbie questions, but can anyone help me out with how to write an interpolation function that can accept borrowed Vecs?  I tried the following, but it doesn't work.  There is some autogenerated docs on crates about Enterpolation implementing Borrow, but it's a bit beyond my decoding ability at this point!  

```
fn approx(xs: &Vec<f64>, ys: &Vec<f64>, newxs: &Vec<f64>) -> Vec<f64> {
    let linear = Linear::builder().elements(ys).knots(xs).build().unwrap();
    linear.sample(newxs).collect()
}
```

**Comment by NicolasKlenert** [comment, marked as answer]
It is a bit confusing to say the least! The reason your code is not working is because `.sample()` does need an iterator or something which can become an iterator with values of `f64`.
A borrowed `Vec` can be converted to an iterator, however this iterator contains borrowed values, that is `&f64` in your case. To get your code working, you have to convert the iterator to contain `f64` values. This is done by the `.copied` method on the iterator itself.

```
fn approx(xs: &Vec<f64>, ys: &Vec<f64>, newxs: &Vec<f64>) -> Vec<f64> {
    let linear = Linear::builder().elements(ys).knots(xs).build().unwrap();
    linear.sample(IntoIterator::into_iter(newxs).copied()).collect()
}
```

I used the verbose version `IntoIterator::into_iter(newxs).copied()` instead of `newxs.into_iter().copied()` as the crate does define its own `into_iter` method. I will look into implementing the sample method in a way such that it also allows borrowed values, it would be nice if it could accept both.

**Reply by jrbrodie77** [reply]
Thank you! Appreciate the prompt answer!
